File: ruby/triangle/triangle.rb
# frozen_string_literal: true

# Implementation of the Triangle exercise in Ruby track on Exercism.
class Triangle
  # @param sides [Array] a list of side lengths of a triangle
  def initialize(sides)
    @sides = sides
    @distinct_side_lengths = sides.uniq.count
    @is_proper_triangle = triangle?
  end

  # @return [Boolean] is the triangle equilateral?
  def equilateral?
    is_proper_triangle && distinct_side_lengths == 1
  end

  # @return [Boolean] is the triangle isoscales?
  def isosceles?
    is_proper_triangle && distinct_side_lengths <= 2
  end

  # @return [Boolean] is the triangle scalene?
  def scalene?
    is_proper_triangle && distinct_side_lengths == 3
  end

  private

  attr_reader :is_proper_triangle, :distinct_side_lengths

  def triangle?
    shortest_side, longest_side = @sides.minmax
    shortest_side.positive? && @sides.sum >= longest_side * 2
  end
end

File: prolog/triangle/triangle.pl
triangle(Side, Side, Side, "equilateral") :- Side > 0.
triangle(Side, Side, Other, "isosceles") :- Sum is Side * 2, Other < Sum.
triangle(Other, Side, Side, "isosceles") :- Sum is Side * 2, Other < Sum.
triangle(Side, Other, Side, "isosceles") :- Sum is Side * 2, Other < Sum.
triangle(Side1, Side2, Side3, "scalene") :-
  Side1 \== Side2,
  Side2 \== Side3,
  Side1 \== Side3,
  Perimiter is Side1 + Side2 + Side3,
  Perimiter > Side1 * 2,
  Perimiter > Side2 * 2,
  Perimiter > Side3 * 2.
